dc/s2n-quic-dc/src/stream/recv/dispatch/queue.rs

```diff
@@ -148,58 +182,120 @@ impl<T> Queue<T> {
+            assert_eq!(control.half, Half::Control);
+        }
 
-    #[inline]
-    pub fn open_receiver(&self) {
-        let Ok(mut inner) = self.lock() else {
-            return;
+        // perform locks in the same order to avoid deadlocks
+        let Ok(mut stream_inner) = self.lock() else {
+            return Err(Closed);
         };
-        trace!("opening receiver");
-        inner.has_receiver = true;
+        let Ok(mut control_inner) = control.lock() else {
+            return Err(Closed);
+        };
+
+        // make sure the stream hasn't been permanently closed
+        ensure!(stream_inner.is_open, Err(Closed));
+        ensure!(control_inner.is_open, Err(Closed));
+
+        debug_assert!(
+            !stream_inner.has_receiver && !control_inner.has_receiver,
+            "receiver already open!\n stream: {stream_inner:?}\ncontrol: {control_inner:?}"
+        );
+
+        stream_inner.has_receiver = true;
+        control_inner.has_receiver = true;
+
+        Ok(())
     }
 
     #[inline]
-    pub fn close_receiver(&self) {
-        let Ok(mut inner) = self.lock() else {
-            return;
+    pub fn close_receiver(&self, control: &Self, half: Half) -> ControlFlow<()> {
+        #[cfg(debug_assertions)]
+        {
+            assert_eq!(self.half, Half::Stream);
+            assert_eq!(control.half, Half::Control);
+        }
+
+        // the Control half owns freeing in the case of poisoning
+        let on_poisoned = if matches!(half, Half::Control) {
+            ControlFlow::Continue(())
+        } else {
+            ControlFlow::Break(())
+        };
+
+        // acquire both locks in the same order to avoid deadlocks or races
+        let Ok(stream_inner) = self.lock() else {
+            return on_poisoned;
         };
-        trace!("closing receiver");
+        let Ok(control_inner) = control.lock() else {
+            return on_poisoned;
+        };
+
+        let (mut inner, other) = match half {
+            Half::Stream => (stream_inner, control_inner),
+            Half::Control => (control_inner, stream_inner),
+        };
+
+        debug_assert!(
+            inner.has_receiver,
+            "receiver already closed:\n{inner:?}\nother: {other:?}"
+        );
+
+        // observe the other half receiver status before dropping the `other` lock
+        let has_other_receiver = other.has_receiver;
+        drop(other);
+
+        let wakers = inner.take_wakers();
         inner.has_receiver = false;
-        inner.app_waker = None;
-        inner.worker_waker = None;
-        inner.queue.clear();
+        // take the queue items out of the lock to avoid mutex poisoning.
+        // note that most of the time this should be empty, which would be a no-op
+        let mut queue = VecDeque::new();
+        queue.append(&mut inner.queue);
+        drop(inner);
+
+        // drop wakers after the lock to avoid potential mutex poisoning
+        wakers.dont_wake();
```

Comment:
Are we waking somewhere else? Otherwise it seems like this will hang a potential read(...) when the stream closes? Or maybe that doesn't actually get hit in practice?

(I think this is not changed in this PR, so not super relevant, but might be another deadlock risk)

Reply:
This function is only called from the receivers themselves so in almost all cases they shouldn't be set. This is mostly for if one of the receivers drops prematurely before getting notified of a new packet.